File: 20. Valid Parentheses/python/main.py
from typing import Dict, List


class Solution:
    def isValid(self, s: str) -> bool:
        stack:List[str] = []
        bracket_map:Dict[str,str] = { "(": ")", "{": "}", "[": "]" }
        for char in s:
            if char in bracket_map:
                stack.append(bracket_map.get(char, ""))
            elif len(stack) ==0 or stack.pop() != char:
                return False

        return len(stack) == 0

File: 20. Valid Parentheses/javascript/main.js
/**
 * @param {string} s
 * @return {boolean}
 */
// Brute force

const isValid1 = function (s) {
  const stack = [];
  const bracketMap = { ")": "(", "}": "{", "]": "[" };
  for (let element of s) {
    if (Object.values(bracketMap).includes(element)) {
      stack.push(element);
    } else if (stack[stack.length - 1] !== bracketMap[element]) return false;
    else stack.pop();
  }
  return stack.length === 0;
};

// Another approach
const isValid2 = function (s) {
  const stack = [];
  const bracketMap = { "(": ")", "{": "}", "[": "]" };

  for (let char of s) {
    if (bracketMap[char]) {
      // If it's an opening bracket, push the corresponding closing bracket
      stack.push(bracketMap[char]);
    } else if (stack.pop() !== char) {
      // If it's a closing bracket, check if it matches the last opened bracket
      return false;
    }
  }

  // If stack is empty, all brackets were properly closed
  return stack.length === 0;
};
